### ISSUE TITLE
Suggestion: allow for manual whitelisting

### ISSUE THREAD

**Issue by jobro13** [pi-hole/pi-hole#336]
For instance, I had to download something behind an adf.ly link - simply replacing adf.ly by the IP doesn't work due CloudFlare - thus I have to fix my own network settings to use the 'default' DNS again. 

The suggestion is: if I surf to adf.ly (or a similiar ad-serving domain), serve a page to temporary whitelist this ad-serving domain. Bonus points if it's for the requesting host only, so I'm not accidentally enabling google ads serving for everyone. After that, return to normal behaviour and serve 404s all the time (except for the adf.ly URI itself)
##

**Comment by PromoFaux** [comment]
Hey could you provide me with a valid adfly link to test, please?

**Comment by PromoFaux** [comment]
Hey, so I've just tested this by whitelisting `cdn.adf.ly` and `adf.ly` and I am able to download files behind adfly pages....


`whitelist.sh adf.ly cdn.adf.ly` should get you what you need....

**Comment by PromoFaux** [comment]
Hi @jobro13 , did this solve the issue for you?

**Comment by schamp** [comment]
+1, I really like the general idea of having the Pi-Hole serve up a page with a control to temporarily whitelist a domain (perhaps with a configurable time window).  I realize this is hugely complex...

**Comment by PromoFaux** [comment]
Ah right, I get it now.

So rather than the page that is returned being blank, you get a page that gives you and option to temporarily whitelist the domain you're trying to access?

At the moment we're working on a secure way of accessing the scripts for whitelisting from a web page. The problem is, the more convienient we make things for the user, the less secure it becomes.  

We have an [open issue](https://github.com/pi-hole/AdminLTE/issues/39) on the Admin page repo that explains just how insecure it would be at the moment.

Once we have something bulletproof, then that is indeed a possibility. But please bear with us :)

**Comment by jobro13** [comment]
Yes, that's right. The main point here is that I don't want access to Pi. Why? Well, I don't want other people in my household sniffing around in Pi (yes, okay, I can create users - but I'm not going to explain mom/grandma how whitelist.sh works!)

I tried searching for these problems but wasn't aware of the other repo. It seems you guys are already on this via the issue provided.

**Comment by DonLexos** [comment]
Problem with this could be that for complete domains/pages this would make sense, but it would also mean that if a page tries to load 10 ads you would get 10 options of whitelisting those domains in the tiny spaces where the ad would be normally.

**Comment by Mcat12** [comment]
And sometimes the ads are loaded in via JS or something that would hide the message. (I have mine set to show "Blocked by Pi-hole" and I only really see it when I go directly to a blocked domain)

**Comment by jobro13** [comment]
I'm not sure if it's clear what I mean. Correct me if I'm wrong, but in most cases ads are loaded via a request to an advertisement domain, e.g: advertisementdomain.com/getadd.php?<options>

But, if I want to temporary whitelist this, I simply go to advertisementdomain.com. Like that, the ad spaces are actually shown as empty 404 pages - but the actual *domain* will be loaded as the temporary whitelist page.

I'm currently running a server - simply going to the IP of my Pi gives me a nice webpage. Also, if I go to any advertisement domain I obviously get this webpage too (due to DNS). (Yes, this is a bad configuration). But I've never seen this webpage popping up so far in any advertisement page.

**Comment by schamp** [comment]
Yes, I am not thinking that I would want this feature so I could load ads from blocked domains, but rather to directly visit a domain that is blocked.  

As a side note, I have seen a number of cases here where the term "whitelist" is overloaded and confusing, I think there are two different kinds of whitelisting that we might want to consider:
* I want to go to "addomain.com" and see content there, but addomain.com is blocked.  By whitelisting addomain.com, I can visit it and load content from that domain from other pages.
* I want to go to favoritewebcomic.com and see content there.  Favoritewebcomic.com is not blocked, but serves ads that are blocked.  I want to support favoritewebcomic.com and unblock the ads referenced on that page, but I don't want to whitelist the domains serving ads for favoritewebcomic.com so that they also show ads on other pages.

In both cases, someone might want a permanent whitelist, or a temporary whitelist.  Granted, I personally don't have strong feelings about the second case.  Also, I don't see a lot of technological ways it could be accomplished using DNS-level ad blocking.  But I think if we could come up with distinct terms for the two cases, we could discuss whitelisting more succinctly and understand better.  Naming is hard, but I hate it when folks say "fetch me a rock" and don't give any parameters, so may I suggest as a starting point something along the lines of "domain whitelisting" for the first cases, and perhaps "ad whitelisting" for the second case?

**Comment by PromoFaux** [comment]
>As a side note, I have seen a number of cases here where the term "whitelist" is overloaded and confusing, I think there are two different kinds of whitelisting that we might want to consider:

Whitelisting, as far as pi-hole is concerned, is all or nothing. If you whitelist an ad serving domain, then it is unblocked for all websites. As we're doing this at a DNS level, we don't have the control to be able to only whitelist ads when on a particular page.

I'm not understanding what is overloaded or confusing about this.

**Comment by PromoFaux** [comment]
I'm going to lock this thread for now. we understand what is being discussed, but at the moment our priorities are stability and more necessary changes.

I'm not saying this is a bad idea, but at the moment, it's a little out of scope!

Thanks for the suggestion.
